### PR TITLE
Swap plugin cache to pickle-able values when done

### LINKED ISSUE #7635: Crash with `-j 2` when linting empty file
### Bug description

Noticed this crash while linting an empty file with `-j 2`. It might be OS dependent as so far I can only reproduce it on MacOS. Linux seems fine.

Bisected the issue to https://github.com/PyCQA/pylint/pull/7284.
In particular this line:
https://github.com/PyCQA/pylint/blob/77e8ae4f3902697babb186482f577a2c8b1a3725/pylint/lint/pylinter.py#L379

/CC: @daogilvie, @DanielNoord

### Configuration

_No response_

### Command used

```shell
pylint -j 2 test.py
```

### Pylint output

```shell
(venv-310) $ pylint -j 2 test.py 
Traceback (most recent call last):
  File "/.../pylint/venv-310/bin/pylint", line 33, in <module>
    sys.exit(load_entry_point('pylint', 'console_scripts', 'pylint')())
  File "/.../pylint/pylint/__init__.py", line 35, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "/.../pylint/pylint/lint/run.py", line 207, in __init__
    linter.check(args)
  File "/.../pylint/pylint/lint/pylinter.py", line 668, in check
    check_parallel(
  File "/.../pylint/pylint/lint/parallel.py", line 141, in check_parallel
    jobs, initializer=initializer, initargs=[dill.dumps(linter)]
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 304, in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 276, in dump
    Pickler(file, protocol, **_kwds).dump(obj)
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 498, in dump
    StockPickler.dump(self, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 487, in dump
    self.save(obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 1375, in save_module
    pickler.save_reduce(_import_module, (obj.__name__,), obj=obj,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 1375, in save_module
    pickler.save_reduce(_import_module, (obj.__name__,), obj=obj,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 1375, in save_module
    pickler.save_reduce(_import_module, (obj.__name__,), obj=obj,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 1375, in save_module
    pickler.save_reduce(_import_module, (obj.__name__,), obj=obj,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 1375, in save_module
    pickler.save_reduce(_import_module, (obj.__name__,), obj=obj,
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 687, in save_reduce
    save(cls)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/.../pylint/venv-310/lib/python3.10/site-packages/dill/_dill.py", line 1439, in save_type
    StockPickler.save_global(pickler, obj, name=name)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/pickle.py", line 1076, in save_global
    raise PicklingError(
_pickle.PicklingError: Can't pickle <class 'astroid.util.Uninferable'>: it's not the same object as astroid.util.Uninferable
```

### Expected behavior

_No crash_

### Pylint version

```shell
pylint 2.16.0-dev
astroid 2.13.0-dev0
Python 3.10.8 (v3.10.8:aaaf517424, Oct 11 2022, 10:14:40) [Clang 13.0.0 (clang-1300.0.29.30)]
```

### OS / Environment

MacOS

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by daogilvie**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

When the dictionary has served its purpose (making plugin loading pre-and-post init a consistent behaviour), we swap it for bools indicating whether or not a module was loaded.
We don't currently use the bools, but it seemed a sensible choice. The main idea is to make the dictionary fully pickle-able, so that when dill pickles the linter for multiprocessing, it doesn't crash horribly.

Closes #7635

### CODE REVIEW

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:407]: You can remove that. The next release of mypy will add better support for `hasattr` at which point we can remove the `type: ignore` as well.

**Review by cdce8p — COMMENTED** [review]
Thanks for the quick fix @daogilvie! Verified that it does resolve the issue. Left a small comment. Besides that the fix is fine IMO. Will leave the actual approval for someone with more experience with multitasking in pylint.

**Review by DanielNoord — COMMENTED** [review]
@daogilvie Have you considered adding a test for this?

**Comment by DanielNoord** [comment]
> I honestly did not, no, because I wasn't sure how to do that beyond what the existing tests are doing in test_check_parallel. I'll look into it this evening; I'm sure there will be a way!

Yeah, I think just linting with `j=2` and a plugin over an empty file might already be enough. Just something to make sure we don't regress on this.
This seems like something that we can easily forget about during a larger refactoring of the multiprocessing code.

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py:407]: I think Marc meant that you can remove the comment explaining the ignore, or at least that would be my suggesiton :smile: 

The comment with the specified code is clear enough in this case as `hasattr` is a known `mypy` issue.

**Comment by daogilvie** [comment]
> > I honestly did not, no, because I wasn't sure how to do that beyond what the existing tests are doing in test_check_parallel. I'll look into it this evening; I'm sure there will be a way!
> 
> Yeah, I think just linting with `j=2` and a plugin over an empty file might already be enough. Just something to make sure we don't regress on this. This seems like something that we can easily forget about during a larger refactoring of the multiprocessing code.

It looks like the reason existing parallel tests didn't catch this is that they _directly register the checkers on the linter_, so they bypass the plugin loading code completely.

I have added a test that doesn't actually lint a file — it just loads the dummy plugin into a linter object and verifies the linter can then be pickled. In addition to not doubling-up the content of other tests, the only time this test would need re-working is if parallel running was revamped so heavily as not to relying on pickling, I think — does it sound like something that will suit?

File: pylint/lint/pylinter.py
Comment (DanielNoord) [inline pylint/lint/pylinter.py:407]:
```suggestion
```

File: tests/test_check_parallel.py
Comment (DanielNoord) [inline tests/test_check_parallel.py:239]:
```suggestion
        """The linter needs to be pickle-able in order to be passed between workers."""
```

**Review by DanielNoord — APPROVED** [review]

File: tests/test_check_parallel.py
Comment (cdce8p) [inline tests/test_check_parallel.py:244]:
```suggestion
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
The test case doesn't seem to work as expected. It still passes when I revert the fix. From the looks of it, the plugin is never loaded - at least `liinter._dynamic_plugins` is empty.

File: tests/test_check_parallel.py
Comment (daogilvie) [reply tests/test_check_parallel.py:244]: I think something has gone weird with my precommit here because I definitely deleted that print call after using it to confirm that the plugin was definitely loading.
I did have some trouble with the hooks and my terminal so perhaps I've ended up committing some awful hybrid. 
Sorry for that, poor performance on my part, I'll review and push up again this morning.

Comment (Pierre-Sassoulas) [reply tests/test_check_parallel.py:244]: I don't know if this was the issue but pre-commit may make changes locally and you need to add those changes before committing again.

Comment (daogilvie) [reply tests/test_check_parallel.py:244]: Yes, I am familiar with it — I use it myself in many projects. It reinstalled the environments twice last night, and hung on one commit, though that might have been the terminal application I was using. I thought I'd come out the other end fine, had no unstaged changes, but didn't think to double-check the actual code I'd committed! Lesson learned in un-deserved confidence there.
In any case, it's back to the drawing board: I fixed what must have gone wrong and now have the test as I originally imagined, but @cdce8p is still spot on that the test is pointless! Test passes even without the fix, modules loaded or no. I'll find another way to test it.

Comment (daogilvie) [reply tests/test_check_parallel.py]: Ok, this was entirely a slip-up on my part — I did not take enough care when checking what I was actually testing. Not the first time I've made that mistake in this codebase (and probably many others 😱). Thanks for your patience as always! This is now resolved.

**Comment by daogilvie** [comment]
I've modified the test to use a non-pickle-safe module explicitly, and to verify the module cannot be pickled (in case it is made pickle-safe in the future independently).
I've also included suggested changes to docstrings and comments, and I've tripled-checked my pre-commit setup and test accuracy 😂

**Comment by daogilvie** [comment]
Ok I am genuinely mortified and confused, sorry. Working on it.

File: pylint/lint/pylinter.py
Comment (daogilvie) [reply pylint/lint/pylinter.py]: Ah, makes sense! Done in the latest commits, thanks both

**Review by DanielNoord — COMMENTED** [review]
I can't test locally but this seems to make sense!

I'll let somebody else test and approve this! Thanks for all your work on this!

File: tests/test_check_parallel.py
Comment (cdce8p) [inline tests/test_check_parallel.py:242]: When testing it locally, this throws a `NotImplementedError` likely because of `wrapt`. Adding the exception should be enough. Just interesting that the CI workflow is green 🤔

> If pickle.dump() or pickle.dumps() is used on an instance of the ObjectProxy class, a NotImplementedError exception is raised, with a message indicating that the object proxy must implement the reduce_ex() method. This is in place of the default TypeError exception with message indicating a pickle error.

https://wrapt.readthedocs.io/en/latest/changes.html#version-1-11-0

**Review by cdce8p — CHANGES_REQUESTED** [review]

File: tests/test_check_parallel.py
Comment (cdce8p) [reply tests/test_check_parallel.py:242]: This only seems to be an issue if astroid is installed in editable mode. Not exactly sure why though. Adding `NotImplementedError` does resolve it.

Comment (daogilvie) [reply tests/test_check_parallel.py:242]: I'll add both, to be sure 😄

Comment (cdce8p) [reply tests/test_check_parallel.py:242]: > I'll add both, to be sure 😄

`PicklingError` is probably enough.
```py
class PicklingError(PickleError):
    """This exception is raised when an unpicklable object is passed to the
    dump() method."""
```

https://github.com/python/cpython/blob/v3.11.0rc2/Lib/pickle.py#L77-L82

**Review by cdce8p — DISMISSED** [review]
Thanks for your work @daogilvie 🚀

**Comment by daogilvie** [comment]
Oh, huh, actually do you mind if I put myself in the `script/.contributors_aliases.json` file before merge? I didn't bother last time but I feel like two issues closed (even if I did _cause_ the second one) means I probably should 😂

**Comment by cdce8p** [comment]
> Oh, huh, actually do you mind if I put myself in the `script/.contributors_aliases.json` file before merge? I didn't bother last time as it felt a bit presumptuous but I feel like two issues closed (even if I did _cause_ the second one) means I probably should 😂

I believe Pierre added you already.
https://github.com/PyCQA/pylint/blob/e8ba3eb002b533e6640d6eeb25a66bab04602951/script/.contributors_aliases.json#L439-L442

**Comment by daogilvie** [comment]
> > Oh, huh, actually do you mind if I put myself in the `script/.contributors_aliases.json` file before merge? I didn't bother last time as it felt a bit presumptuous but I feel like two issues closed (even if I did _cause_ the second one) means I probably should 😂
> 
> I believe Pierre added you already.
> 
> https://github.com/PyCQA/pylint/blob/e8ba3eb002b533e6640d6eeb25a66bab04602951/script/.contributors_aliases.json#L439-L442

Oh that's great! Thank you again @Pierre-Sassoulas for covering for my laziness 👀

**Comment by cdce8p** [comment]
> Oh that's great! Thank you again @Pierre-Sassoulas for covering for my laziness 👀 — I might just shorten my name though if that's ok 😄

Please do if you like!

**Review by cdce8p — APPROVED** [review]

**Pull request merged**
